### ISSUE TITLE
Support for Mastodon Android app

### ISSUE THREAD

**Issue by kr428** [friendica/friendica#11415]
The upcoming Mastodon Android app does list Friendica instances but apparently doesn't allow for connecting to them, resulting in a "missing parameter" error, see
https://github.com/mastodon/mastodon-android/issues/23
Is there a way to fix this? Or is it too specific to Mastodon to adapt Friendica API to these calls here?

**Comment by annando** [comment]
I currently don't have access to the Android App, it doesn't seem to be available at the playstore at the moment. I will perform some tests when I can install it on my system.

**Comment by kr428** [comment]
It's available in this fdroid repo by now:
https://apt.izzysoft.de/fdroid/index.php

They seem to provide the version built for 1.0 that is in review for Play Store at the moment.

**Comment by annando** [comment]
I don't know that repository. I do trust F-Droid and I have to trust the Play Store.

**Comment by kr428** [comment]
You might use this binary then...?
https://github.com/mastodon/mastodon-android/releases

**Comment by annando** [comment]
I now installed it. It seems to work with the develop branch.

**Comment by kr428** [comment]
Ok, so there seems a chance it could work with a release at some point...?

**Comment by annando** [comment]
In the next days I will dig deeper into possible issues. Surely the next release will contain everything.

**Comment by ne20002** [comment]
I tried yesterday and couldn't log in to existing account. Error message is something about malformed json. I really like to give the app a try.

**Comment by ghost** [comment]
I thing this is related to to the login issue in the Mastodon App

`index [WARNING]: Incomplete request data {"request":{"pagename":"oauth/token"}} - {"file":"Token.php","line":71,"function":"post","uid":"08c591","process_id":419055}`

**Comment by foss-** [comment]
Sign-in was possible using Friendica, version 2022.09-rc that is running at the web location https://squeet.me. The database version is 1484/1484, the post update version is 1484/1484.
Mastodon 1.1.3 from F-Droid

**Comment by kr428** [comment]
Still doesn't work for loma.ml (on 2022-10), trying to log in from current (1.1.3) Mastodon app still fails with "missing parameters" after selecting the instance.

**Comment by ne20002** [comment]
I removed all my connected apps in Friendica and logged out on my devices. I can't get logged in again. It's always telling me that **the app can't retrieve an OAuth token**.
Happens with Tusky and FediLab on both my mobile devices. 
Did try a few checks on my side: disable modsecurity, disabled caching, restarted.

I may need to recheck my whole installation. Or set up a new instance (with more usefull name right away).

**Comment by kr428** [comment]
Current state as far as I tested ("official" Mastodon app as well as Megalodon which seems a fork of it):

- venera.social : "missing parameters" error in the app, no redirect to login screen.
- loma.ml : "missing parameters" error in the app, no redirect to login screen.
- squeet.me : redirect to login apparently works (don't have an account there, but behaviour is different)

Mentioning @loma-one here. It _somehow_ seems an infrastructure issue then?

**Comment by foss-** [comment]
Since login works on squeet me and pirati ca this seems instance related. Maybe the title should be updated accordingly since this is not a general problem.

**Comment by annando** [comment]
I recommend that the admin of these servers should perform a test on their machine. Possibly the logs (especially the webserver log) can tell something.

**Comment by loma-one** [comment]
apache, php & mysql log return no results. With the Apple APP the login is possible. 

For a short time I disabled all `<IfModule mod_headers.c>` and tried to log in with the Android App. Error as described above.

When I know what to look for, I'll take another closer look. 

![loma_m_mastodonApp](https://user-images.githubusercontent.com/44441246/215049537-7b633479-a72c-47b4-8113-55c7db3d6541.png)

**Comment by kr428** [comment]
> apache, php & mysql log return no results. With the Apple APP the login is possible.
> 
> For a short time I disabled all `<IfModule mod_headers.c>` and tried to log in with the Android App. Error as described above.
> 
@loma-one Thanks for looking into this. Just one thing that comes to mind: When I try to connect to loma.ml using Subway Tooter, I (also) can't get to the authentication page at all, the client fails out early with an HTTP 401 saying  "can't get server information, unauthorized. This API requires login"
Assuming these clients altogether make use of the same API: Is it possible that there's some dedicated protection in place for whichever endpoint is used at this stage to retrieve server information?

**Comment by loma-one** [comment]
The firewall is listening on port 80; 443. Which port must be enabled here, if any? Which API endpoint is required here?

**Comment by loma-one** [comment]
Which API endpoint is involved? 
There are no messages about rejected logins in the logs.

---
squeet.me returns the same error when trying to log in via Subway Tooter. 
![Screenshot_20230128_085805_jp juggler subwaytooter](https://user-images.githubusercontent.com/44441246/215254729-384e8b60-a606-445d-9388-e7aaa7429b4f.jpg)

**Comment by kr428** [comment]
No real idea, didn't manage to find out so far. I _thought_ about https://loma.ml/api/v1/instance at first but that seems to work at least in the browser. Do you have any 401's in the access log if you try to sign in using one of these apps? Shouldn't be a login error (yet) I guess as it happens quite before that... .

**Comment by loma-one** [comment]
no entries that refer to it

**Comment by kr428** [comment]
Hmmm... doesn't apache log a 401? 🙁 Needs more debugging I guess, I'll see to have a look at the app code in a spare moment...

**Comment by kr428** [comment]
(I'm a bit confused about that. Not sure if that's related, but I tried linking to a public friendica post of mine within Facebook some days ago, and Facebook tries to generate a preview which also just resulted in an apache 401 page while clicking the link even in an anonymous browser would perfectly display the content of the post....)

**Comment by kr428** [comment]
@loma-one Trying to dive a bit deeper into that with a Friendica stable (on uberspace), I see the connection procedure using Mastodon or Megalodon causes these three requests:

GET /api/v1/instance     - after entering the instance name, used to retrieve basic instance information
POST /api/v1/apps        - after trying to continue from there (seems most likely an unauthenticated call at this stage?)
GET /oauth/authorize    - for triggering oauth procedure.

Looking at where this fails trying to connect to loma.ml, it _seems_ the POST request is the culprit at this point. Trying this with my uberspace instance, this procedure works and gets me to the oauth login screen. Maybe you're in some way blocking / restricting the POST to /api/v1/apps at this stage...?
(Edit: The Subway Tooter issue seems unrelated; apparently in this case a POST /api/meta is tried and fails.=

**Comment by foss-** [comment]
> squeet.me returns the same error when trying to log in via Subway Tooter.

@loma-one While I can confirm this problem on subway tooter, could you please retry with Fedilab or Tusky both of which seem a bit more mature than Subway Tooter and are known to work with Friendica.

@kr428 could you adjust the title of this bug to make it more descriptive since currently it is rather misleading.

**Comment by kr428** [comment]
> @kr428 could you adjust the title of this bug to make it more descriptive since currently it is rather misleading.

Why so? Can do but the ticket actually was about supporting the "Mastodon" app (and now its forks such as Megalodon).

**Comment by foss-** [comment]
Yeah but this is as stated already not a general friendica issue - at least I am unable to reproduce this problem on Fedilab or Tusky using squeet me or pirati ca.

Do you reproduce this problem on either client with one of the two instances?

**Comment by kr428** [comment]
Fedilab and Tusky work well on all instances I tested them. Mastodon or Megalodon don't, at least on some instances, that's what this issue is about. As they work with Mastodon services out of the box, it must be either an issue of Friendicas implementation of the Mastodon client API or (which seems more likely) an administration issue on particular instances. Would be interesting to know the *exact* cause though.

**Comment by foss-** [comment]
Still no repro retesting with joinfriendica.de account.
Version: This is Friendica, version 2025.07-rc that is running at the web location https://joinfriendica.de. The database version is 1580/1580, the post update version is 1550/1550.

- login into friendica instance in default browser on Android
- tap Log-In in mastodon app
- manually enter friendica server
- Authorise application connection > Yes
- back in mastodon app in logged in state

Is this issue persisting for you with 2025.07RC @kr428 ?

**Comment by kr428** [comment]
> Still no repro retesting with joinfriendica.de account. Version: This is Friendica, version 2025.07-rc that is running at the web location https://joinfriendica.de. The database version is 1580/1580, the post update version is 1550/1550.
> 
> Is this issue persisting for you with 2025.07RC [@kr428](https://github.com/kr428) ?

No. I'll close this off. Seems this has already been working with the 2024.12 but I haven't used these apps that often recently.
